### PR TITLE
Handle hash instead of star as invalid changelog line; [WIP]

### PR DESCRIPTION
**Pull request by Antondomashnev**
This PR addresses the issue #16 

@dblock I have doubts about the method name I've chosen to check if the changelog line almost looks like changelog line 😄 
```ruby
Danger::Changelog::ChangelogLine.changelog_line?(line)
```

I also have one idea how to refactor the code a bit:  
If the `Danger::Changelog::ChangelogLine` can be initialized with a `line` and that it can have a type: `["you_contribution", "header", "changelog"]` and depending on the `type`, it can be validated properly. So eventually it moves all logic about validation inside one class `Danger::Changelog::ChangelogLine`, instead of two currently. @dblock what do you think?

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ### 0.2.0 (Next)
 
 * [#13](https://github.com/dblock/danger-changelog/pull/13): Fix: suggested CHANGELOG entry may not be of correct format - [@dblock](https://github.com/dblock).
+* [#17](https://github.com/dblock/danger-changelog/pull/17): Fix: improve CHANGELOG line validation - [@antondomashnev](https://github.com/antondomashnev).
```
Comment (dblock) [inline CHANGELOG.md]: Maybe something more explicit like "handle changelog lines that mistakenly start with a symbol other than *"?

**Comment by dblock** [comment]
Definitely think that the current check is duplicating code, so I would like to refactor this further before merging. 

Initializing a `ChangelogLine` with `line` makes total sense. It should be able to say whether this is a changelog line and whether it's a valid one. It's possible that the approach to try and parse the line by extracting all its parts and then recombine them and see whether they all fell in the same place could work, or maybe we need some kind of "distance" between a valid line and what we have, so basically anything that looks almost like a changelog line is good to be checked? Give it a try!

File: CHANGELOG.md (hunk 1 above)
Comment (Antondomashnev) [inline CHANGELOG.md]: Yes, that sounds better and clearer. Will update it after implementing some refactoring.

**Comment by Antondomashnev** [comment]
@dblock Will work on some refactoring, starting with the `ChangelogLine` with `Line` initialization and parsing inside.
I think the order validation should be a separate feature, to not mess up tasks and PRs 😉

File: lib/changelog/changelog_file.rb [hunk 2]
```diff
@@ -38,15 +38,21 @@ def parse
         @your_contribution_here = false
         @bad_lines = []
         File.open(filename).each_line do |line|
-          # ignore lines that aren't changes
-          next unless line[0] == '*'
+          # ignore new line
+          next if line == "\n"
```
Comment (dblock) [inline lib/changelog/changelog_file.rb]: I think the line should be stripped, then skipped if it's blank.

File: lib/changelog/changelog_file.rb [hunk 3]
```diff
@@ -38,15 +38,21 @@ def parse
...
+
+          changelog_line = ChangelogLineParser.parse(line)
+
+          if changelog_line.nil? || !changelog_line.valid?
```
Comment (dblock) [inline lib/changelog/changelog_file.rb]: Add `.invalid?` as the opposite of `valid?` and it will make this more logical.

**Comment by Antondomashnev** [comment]
@dblock thanks for the ongoing review 👍 Hope will make it finished by the end of the week. But yes, eventually, I think this refactoring could bring lots of new validation possibilities.

**Comment by dblock** [comment]
LMK when this is in a good state and I'll go through it thoroughly.

File: lib/changelog/changelog_file.rb (hunk 2 above)
Comment (Antondomashnev) [inline lib/changelog/changelog_file.rb]: I've changed the line to this `next if line.strip.empty?`, seems fine to me. For you?

**Comment by Antondomashnev** [comment]
@dblock you can check now. Will clean up the history after review 👍

**Comment by dblock** [comment]
I've squash-merged this, thanks for your excellent work.

Want to help me maintain this gem? Make a release? What's your rubygems email?

**Comment by Antondomashnev** [comment]
@dblock thanks for merging 👍 
Yes, would love to help you with it, the email is antondomashnev[at]gmail.com.

**Comment by dblock** [comment]
@Antondomashnev You have all access now. Maybe start by committing a RELEASING.md like https://github.com/slack-ruby/slack-ruby-client/blob/master/RELEASING.md and following those instructions?

**Comment by Antondomashnev** [comment]
@dblock that sounds like a plan. Also, want to think about additional rules we can use for the changelog. Such as line orders or versions check.

**Comment by dblock** [comment]
Cool. For any ideas take the habit of opening a separate feature request. This project is great for first time contributors - lots of bite-sized work!

**Comment by dblock** [comment]
Oh and please do keep making pull requests for your own changes to get code reviews. Feel free to merge other people's of course.

**Pull request merged**